torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,273 @@
+    strides = at::detail::defaultStrides(sizes);
+  }
+  return std::make_pair(sizes, strides);
+}
+
+Node* insertAllocStorageNode(
+    std::shared_ptr<Graph>& graph,
+    uint64_t total_size) {
+  auto* storage = graph->create(prim::AllocateStorage, 1);
+  storage->i_(attr::total_size, total_size);
+
+  auto deviceType = jit::tensorexpr::pickDeviceType(graph);
+  if (deviceType.has_value()) {
+    storage->i_(attr::device, static_cast<int8_t>(deviceType.value().type()));
+  } else {
+    storage->i_(attr::device, static_cast<int8_t>(at::kCPU));
+  }
+  storage->insertBefore(graph->nodes().front());
+  return storage;
+}
+
+void insertAllocTensorNodes(
+    std::shared_ptr<Graph>& graph,
+    Node* storage,
+    std::unordered_map<LiveRange, Region, live_range_hash> allocations,
+    std::map<LiveRange, const Value*, live_range_start_cmp>
+        manage_range_values) {
+  uint64_t total_size = storage->i(attr::total_size);
+  for (auto& item : manage_range_values) {
+    auto lvr = item.first;
+    auto region = allocations[lvr];
+    auto allocation = item.second;
+
+    // const_cast fishy?
+    auto node = const_cast<Node*>(allocation->node());
+
+    // the way that this node magically *becomes* the out varaint is simply
+    // by add an extra input. this is because op resolution happens
+    // at runtime via the op registry (by matching on the schema).
+    auto* alloc = graph->create(prim::AllocateTensor, 1);
+    node->addInput(alloc->output());
+    GRAPH_DEBUG("inserting allocation op for ", node->getOperator().schema());
+    alloc->insertBefore(node);
+    alloc->addInput(storage->output());
+
+    auto ttp = allocation->type()->expect<c10::TensorType>();
+    std::vector<int64_t> sizes, strides;
+    std::tie(sizes, strides) = getSizesStrides(ttp);
+    TORCH_CHECK(
+        region.offset + region.size <= total_size,
+        "trying to create an allocation that exceeds previously planned memory");
+    alloc->i_(attr::size, region.size);
+    alloc->i_(attr::offset, region.offset);
+    alloc->is_(attr::sizes, sizes);
+    alloc->is_(attr::stride, strides);
+    alloc->i_(attr::device, static_cast<int8_t>(storage->i(attr::device)));
+    alloc->i_(attr::dtype, static_cast<int8_t>(ttp->scalarType().value()));
+  }
+}
+
+bool hasOutVariant(Node* node) {
+  for (const auto& variant : getAllOperatorsFor(node->kind())) {
+    auto variant_args = variant->schema().arguments();
+    /* TODO
+      aten::cat.names_out(Tensor[] tensors, str dim, *, Tensor(a!) out) ->
+      (Tensor(a!)) aten::cat.out(Tensor[] tensors, int dim=0, *,
+      Tensor(a!) out) -> (Tensor(a!))
+    */
+    auto maybe_out_arg =
+        std::find_if(variant_args.begin(), variant_args.end(), [](auto arg) {
+          return arg.name() == "out";
+        });
+    if (maybe_out_arg != variant_args.end()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+std::pair<std::vector<const Node*>, std::unordered_map<const Value*, uint64_t>>
+getManagedValues(
```

Comment:
could you add a sentence or two description above the function what this is achieving?